cider/function/aggregate/CiderRuntimeFunctions.h

```diff
@@ -109,6 +109,67 @@ ALWAYS_INLINE void cider_agg_id(T& agg_val, const T& val) {
 DEF_CIDER_SIMPLE_AGG_FUNCS(id, cider_agg_id)
 /********************************************************************************/
 
+/********************** Project Id Functions *****************************/
+
```

Comment:
Are these `Project Id Functions` only used in agg case? What will be the actual generated IR like?

Reply:
Not exactly, Project Id Functions are specialized for pure project output.

Reply:
// select col, -col from table;
```llvm
; Function Attrs: uwtable
define void @query_group_by_template(i8** nocapture readnone %byte_stream, i8* nocapture readonly %literals, i64* nocapture readnone %row_count_ptr, i64* nocapture readonly %frag_row_off_ptr, i32* %max_matched_ptr, i64* %agg_init_val, i64** %group_by_buffers, i32 %frag_idx, i64* %join_hash_tables, i32* noalias %total_matched, i32* %error_code) #24 {
.entry:
  %0 = getelementptr i8*, i8** %byte_stream, i32 0, !NativeCodegen.cpp !20
  %1 = load i8*, i8** %0, !NativeCodegen.cpp !20
  %row_count = load i64, i64* %row_count_ptr, align 8, !NativeCodegen.cpp !20
  %max_matched = load i32, i32* %max_matched_ptr, align 8, !NativeCodegen.cpp !20
  %crt_matched = alloca i32, !NativeCodegen.cpp !20
  %old_total_matched = alloca i32, !NativeCodegen.cpp !20
  %2 = call i32 @pos_start_impl(i32* %error_code), !NativeCodegen.cpp !20
  %3 = call i32 @pos_step_impl(), !NativeCodegen.cpp !20
  %4 = call i32 @group_buff_idx_impl(), !NativeCodegen.cpp !20
  %5 = sext i32 %2 to i64, !NativeCodegen.cpp !20
  %6 = getelementptr i64*, i64** %group_by_buffers, i32 %4, !NativeCodegen.cpp !20
  %col_buffer = load i64*, i64** %6, align 8, !NativeCodegen.cpp !20
  %result_buffer = call i64* @init_shared_mem_nop(i64* %col_buffer, i32 0), !NativeCodegen.cpp !20
  %7 = icmp slt i64 %5, %row_count, !NativeCodegen.cpp !20
  br i1 %7, label %.loop.preheader, label %.exit, !NativeCodegen.cpp !20

.loop.preheader:                                  ; preds = %.entry
  %8 = sext i32 %3 to i64, !NativeCodegen.cpp !20
  br label %.forbody, !NativeCodegen.cpp !20

.forbody:                                         ; preds = %.forbody, %.loop.preheader
  %pos = phi i64 [ %5, %.loop.preheader ], [ %10, %.forbody ], !NativeCodegen.cpp !20
  %9 = call i32 @row_func(i64* %result_buffer, i64* null, i32* %crt_matched, i32* %total_matched, i32* %old_total_matched, i32 %max_matched, i64* %agg_init_val, i64 %pos, i64* %frag_row_off_ptr, i64* %row_count_ptr, i8* %literals, i8* %1, i64* %join_hash_tables), !NativeCodegen.cpp !20
  %10 = add i64 %pos, %8, !NativeCodegen.cpp !20
  %11 = icmp slt i64 %10, %row_count, !NativeCodegen.cpp !20
  br i1 %11, label %.forbody, label %._crit_edge, !NativeCodegen.cpp !20

._crit_edge:                                      ; preds = %.forbody
  br label %.exit, !NativeCodegen.cpp !20

.exit:                                            ; preds = %._crit_edge, %.entry
  call void @write_back_nop(i64* %col_buffer, i64* %result_buffer, i32 0), !NativeCodegen.cpp !20
  ret void, !NativeCodegen.cpp !20
}

; Function Attrs: alwaysinline
define i32 @row_func(i64* %group_by_buff, i64* %varlen_output_buff, i32* %crt_matched, i32* %total_matched, i32* %old_total_matched, i32 %max_matched, i64* %agg_init_val, i64 %pos, i64* %frag_row_off, i64* %num_rows_per_scan, i8* %literals, i8* %col_buf0, i64* %join_hash_tables) #25 {
entry:
  br i1 true, label %filter_true, label %filter_false, !DiamondCodegen.cpp !22

filter_true:                                      ; preds = %entry
  %0 = load i32, i32* %total_matched, !GroupByAndAggregate.cpp !23
  %1 = sext i32 %0 to i64, !GroupByAndAggregate.cpp !23
  %2 = add i32 %0, 1, !GroupByAndAggregate.cpp !23
  store i32 %2, i32* %total_matched, !GroupByAndAggregate.cpp !23
  %3 = icmp ne i64* %group_by_buff, null, !GroupByAndAggregate.cpp !23
  br i1 %3, label %groupby_nullcheck_true, label %groupby_nullcheck_false, !DiamondCodegen.cpp !22

filter_false:                                     ; preds = %groupby_nullcheck_true, %entry
  ret i32 0, !GroupByAndAggregate.cpp !23

groupby_nullcheck_true:                           ; preds = %filter_true
  %4 = call i8* @cider_ColDecoder_extractArrowBuffersAt(i8* %col_buf0, i64 1), !ColumnIR.cpp !24
  %5 = call i64 @fixed_width_int_decode(i8* %4, i32 1, i64 %pos), !ColumnIR.cpp !24
  %6 = trunc i64 %5 to i8, !ColumnIR.cpp !24
  %7 = getelementptr i64, i64* %group_by_buff, i32 0, !TargetExprBuilder.cpp !15
  %8 = load i64, i64* %7, !TargetExprBuilder.cpp !15
  %9 = inttoptr i64 %8 to i8*, !TargetExprBuilder.cpp !15
  %10 = call i8* @cider_ColDecoder_extractArrowBuffersAt(i8* %9, i64 1), !TargetExprBuilder.cpp !15
  call void @cider_agg_id_proj_int8(i8* %10, i64 %1, i8 %6), !CiderAggregateCodeGenerator.cpp !21
  %11 = sub i8 0, %6, !ArithmeticIR.cpp !25
  %12 = getelementptr i64, i64* %group_by_buff, i32 1, !TargetExprBuilder.cpp !15
  %13 = load i64, i64* %12, !TargetExprBuilder.cpp !15
  %14 = inttoptr i64 %13 to i8*, !TargetExprBuilder.cpp !15
  %15 = call i8* @cider_ColDecoder_extractArrowBuffersAt(i8* %14, i64 1), !TargetExprBuilder.cpp !15
  call void @cider_agg_id_proj_int8(i8* %15, i64 %1, i8 %11), !CiderAggregateCodeGenerator.cpp !21
  br label %filter_false, !DiamondCodegen.cpp !26

groupby_nullcheck_false:                          ; preds = %filter_true
  ret i32 -1, !GroupByAndAggregate.cpp !23
}

End of IR
```